Updated Auth Getting started Guide (#25824)
backstage/backstage: docs/getting-started/config/authentication.md

> 

Reviewer: Outside of the diff, but can we change the description to `How to setup authentication for your Backstage app`. Both grammar and being consistent about Backstage app as the a single instance of Backstage customized for your company.

Author: Udpated

> ### Setting up authentication

Reviewer: Can we give a little context on what we're doing before jumping in? It's been a minute since I did the doc rewrites, but I think I was trying to do a `Summary` section for each doc that outlines what you'll learn after reading it. 

Something like,
```
### Summary
We'll be walking you through how to setup authentication for your Backstage app using Github. After finishing this guide, you'll have both working authentication and users in your Backstage app to match to the users logging in!
```

Author: Done 👍

There are multiple authentication providers available for you to use with Backstage, feel free to follow [the instructions for adding authentication](../../auth/index.md).

For this tutorial we choose to use GitHub, a free service most of you might be familiar with. For other options, see
[the auth provider documentation](../../auth/github/provider.md#create-an-oauth-app-on-github).

Go to [https://github.com/settings/applications/new](https://github.com/settings/applications/new) to create your OAuth App. The `Homepage URL` should point to Backstage's frontend, in our tutorial it would be `http://localhost:3000`. The `Authorization callback URL` will point to the auth backend, which will most likely be `http://localhost:7007/api/auth/github/handler/frame`.

![Screenshot of the GitHub OAuth creation page](../../assets/getting-started/gh-oauth.png)

> Take note of the `Client ID` and the `Client Secret` (clicking the "Generate a new client secret" button will get this value for you). Open `app-config.yaml`, and add your `clientId` and `clientSecret` to this file. It should end up looking like this:

Reviewer: ```suggestion
Take note of the `Client ID` and the `Client Secret` (clicking the "Generate a new client secret" button will get this value for you). Open `app-config.yaml`, and add them as `clientId` and `clientSecret` in this file. It should end up looking like this:
```

```yaml title="app-config.yaml"
auth:
  # see https://backstage.io/docs/auth/ to learn about auth providers
  /* highlight-add-start */
  environment: development
  /* highlight-add-end */
  providers:
    # See https://backstage.io/docs/auth/guest/provider
    guest: {}
    /* highlight-add-start */
    github:
      development:
        clientId: YOUR CLIENT ID
        clientSecret: YOUR CLIENT SECRET

>         signIn:

Author: Sure, made some changes that should address this can you give it a look, please?

          resolvers:
            # Matches the GitHub username with the Backstage user entity name.
            # See https://backstage.io/docs/auth/github/provider#resolvers for more resolvers.
            - resolver: usernameMatchingUserEntityName
    /* highlight-add-end */
```

:::note Note

The default Backstage app comes with a guest Sign In Resolver. This resolver makes all users share a single "guest" identity and is only intended as a minimum requirement to quickly get up and running. You can read more about how [Sign In Resolvers](../../auth/identity-resolver.md#sign-in-resolvers) play a role in creating a [Backstage User Identity](../../auth/identity-resolver.md#backstage-user-identity) for logged in users.

> :::

Reviewer: Let's move this up top, seems like an important detail before diving into the full guide.

Author: Sure, done!

### Add sign-in option to the frontend

> Backstage will re-read the configuration. If there's no errors, that's great! We can continue with the next part of the configuration. The next step is needed to change the sign-in page, this you actually need to add in the source code.

Reviewer: This is probably ancient text, can we add FAQs for what to do if you get errors?

Author: I've just removed it.

> Backstage will re-read the configuration. If there's no errors, that's great! We can continue with the next part of the configuration. The next step is needed to change the sign-in page, this you actually need to add in the source code.

Reviewer: Also, 
```suggestion
Backstage will re-read the configuration. If there's no errors, that's great! We can continue with the next part of the configuration. The next step is needed to change the sign-in page. For this, you'll actually need to write some code.
```

},
```

### Add the auth provider to the backend

> To add the auth provider to the backend we will first need to install the package by running this command:

Reviewer: ```suggestion
To add the auth provider to the backend, we will first need to install the package by running this command:
```

```bash title="from your Backstage root directory"
yarn --cwd packages/backend add @backstage/plugin-auth-backend-module-github-provider
```

Then we will need to add this line:

```ts title="in packages/backend/src/index.ts"
backend.add(import('@backstage/plugin-auth-backend'));
/* highlight-add-start */
backend.add(import('@backstage/plugin-auth-backend-module-github-provider'));
/* highlight-add-end */
```

> Restart Backstage from the terminal, by stopping it with `Control-C`, and starting it with `yarn dev`. You should be welcomed by a login prompt!

Reviewer: ```suggestion
Restart Backstage from the terminal, by stopping it with `Ctrl+C`, and starting it with `yarn dev`. You should be welcomed by a login prompt!
```

Bit of a nit, but `Control` as a whole word is painful to read lol.

:::note Note

Sometimes the frontend starts before the backend resulting in errors on the sign in page. Wait for the backend to start and then reload Backstage to proceed.

:::

> If you try to login at this point you will get a "Failed to sign-in, unable to resolve user identity" message, read on as we'll fix that next.

Reviewer: ```suggestion
If you try to login at this point you will get a `Failed to sign-in, unable to resolve user identity` message, read on as we'll fix that next.
```

Author: I prefer not to use backticks for things that are not code.

> If you try to login at this point you will get a "Failed to sign-in, unable to resolve user identity" message, read on as we'll fix that next.

Reviewer: As that guy that doesn't read all of the instructions before jumping in, can we move this up a bit higher? If you're me, you won't read this until after you've spent 20 mins debugging.

Author: Moved it up before the note.

The recommended approach for adding Users, and Groups, into your Catalog is to use one of the existing Org Entity Providers - [like this one for GitHub](https://backstage.io/docs/integrations/github/org) - or if those don't work you may need to [create one](https://backstage.io/docs/features/software-catalog/external-integrations#custom-entity-providers) that fits your Organization's needs.

For the sake of this guide we'll simply step you though adding a User to the `org.yaml` file that is included when you create a new Backstage instance. Let's do that:

1. Fist open the `/examples/org.yaml` file in your text editor of choice
2. At the bottom we'll add the following YAML:

   ```yaml
   ---
   apiVersion: backstage.io/v1alpha1
   kind: User
   metadata:
     name: YOUR GITHUB USERNAME
   spec:
     memberOf: [guests]
   ```

3. Now make sure to replace the text "YOUR GITHUB USERNAME" with your actual GitHub User name.

> Let's restart Backstage from the terminal ponce more, by stopping it with `Control-C`, and starting it with `yarn dev`. You should now be able to log into Backstage and see items in you Catalog.

Reviewer: ```suggestion
Let's restart Backstage from the terminal once more, by stopping it with `Ctrl+C`, and starting it with `yarn dev`. You should now be able to log into Backstage and see items in your Catalog.
```

> Let's restart Backstage from the terminal ponce more, by stopping it with `Control-C`, and starting it with `yarn dev`. You should now be able to log into Backstage and see items in you Catalog.

Reviewer: I think this can also now be done by just pressing enter in the terminal that Backstage is running in?

Author: As I'm not sure I'd like to just leave it.

- [Using organizational data from GitHub](../../integrations/github/org.md)

## Setting up a GitHub Integration

The GitHub integration supports loading catalog entities from GitHub or GitHub Enterprise. Entities can be added to static catalog configuration, registered with the catalog-import plugin, or discovered from a GitHub organization. Users and Groups can also be loaded from an organization. While using [GitHub Apps](../../integrations/github/github-apps.md) might be the best way to set up integrations, for this tutorial you'll use a Personal Access Token.

Create your Personal Access Token by opening [the GitHub token creation page](https://github.com/settings/tokens/new). Use a name to identify this token and put it in the notes field. Choose a number of days for expiration. If you have a hard time picking a number, we suggest to go for 7 days, it's a lucky number.

![Screenshot of the GitHub Personal Access Token creation page](../../assets/getting-started/gh-pat.png)

Set the scope to your likings. For this tutorial, selecting `repo` and `workflow` is required as the scaffolding job in this guide configures a GitHub actions workflow for the newly created project.

> For this tutorial, we will be writing the token to `app-config.local.yaml`. This file might not exist for you, so if it doesn't go ahead and create it alongside the `app-config.yaml` at the root of the project. This file should also be excluded in `.gitignore`, to avoid accidental committing of this file.

Author: I've added a link but we don't have anything that's really explicit about this specific file. Perhaps that could be added in a follow up PR?
